tests/test_checkpointing.py

```diff
@@ -0,0 +1,38 @@
+import pytest
+import firedrake as fd
+import fireshape as fs
+
+
+@pytest.mark.parametrize("controlspace_t", [fs.FeControlSpace,
+                                            fs.FeMultiGridControlSpace,
+                                            fs.BsplineControlSpace])
+def test_checkpointing(controlspace_t):
+    mesh = fd.UnitSquareMesh(5, 5)
+
+    if controlspace_t == fs.BsplineControlSpace:
+        bbox = [(-1, 2), (-1, 2)]
+        orders = [2, 2]
+        levels = [4, 4]
+        Q = fs.BsplineControlSpace(mesh, bbox, orders, levels)
+    elif controlspace_t == fs.FeMultiGridControlSpace:
+        Q = fs.FeMultiGridControlSpace(mesh, refinements=1, order=2)
+    else:
+        Q = controlspace_t(mesh)
+
+    inner = fs.LaplaceInnerProduct(Q)
+
+    q = fs.ControlVector(Q, inner)
+    p = fs.ControlVector(Q, inner)
+
+    from firedrake.petsc import PETSc
+    rand = PETSc.Random().create(mesh.comm)
```

Comment:
Which random numbers are these? If possible, I suggest uniform random numbers in the interval [1,2], and then
`assert q.norm() > 1` in line 35 (so that we are sure that we never test we zero vectors)